cranelift-filetests/src/function_runner.rs

```diff
@@ -1,53 +1,301 @@
+}
+
+impl Display for Operator {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        f.write_str(match self {
+            Self::EQ => "==",
+            Self::NEQ => "!=",
+        })
+    }
+}
+
+impl FromStr for Operator {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "==" => Ok(Self::EQ),
+            "!=" => Ok(Self::NEQ),
+            _ => Err(format!("Unsupported operator {}", s)),
+        }
+    }
+}
+
+macro_rules! try_into_int {
+    ($type: ty, $name: ident) => {
+        struct $name {}
+        impl $name {
+            fn try_into(d: ConstantData) -> Result<$type, String> {
+                if d.len() == (mem::size_of::<$type>()) {
+                    let v = d.into_vec().into_iter().rev();
+                    let mut r: $type = 0;
+                    for b in v {
+                        let (mut shifted, _) = r.overflowing_shl(8);
+                        shifted |= b as $type;
+                        r = shifted;
+                    }
+                    Ok(r)
+                } else {
+                    Err(format!(
+                        "Incorrect vector size: {}, expected {}",
+                        d.len(),
+                        mem::size_of::<$type>()
+                    ))
+                }
+            }
+        }
+    };
+}
+
+try_into_int!(u8, ConstantDataU8);
+try_into_int!(u16, ConstantDataU16);
+try_into_int!(u32, ConstantDataU32);
+try_into_int!(u64, ConstantDataU64);
+try_into_int!(u128, ConstantDataU128);
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Value {
+    Bool(bool),
+    F32(f32),
+    F64(f64),
+    I8(u8),
+    I16(u16),
+    I32(u32),
+    I64(u64),
+    I128(u128),
+}
+
+impl Value {
+    fn parse_as(s: &str, ty: types::Type) -> Result<Self, String> {
+        if ty.is_vector() {
+            return Err(format!("Expected a non-vector type, not {}", ty));
+        }
+
+        macro_rules! consume {
+            ( $ty:ident, $match_fn:expr ) => {{
+                ConstantData::default().append($match_fn)
+            }};
+        }
+
+        fn boolean_to_vec(value: bool) -> Vec<u8> {
+            let mut buffer = vec![0; 1];
+            buffer[0] = if value { 1 } else { 0 };
+            buffer
+        }
+
+        let mut p = Parser::new(s);
+
+        let constant_data = match ty {
+            types::I8 => {
+                let data = p.match_constant_data(ty).unwrap();
+                Self::I8(ConstantDataU8::try_into(data).unwrap())
+            }
+            types::I16 => {
+                let data = p.match_constant_data(ty).unwrap();
+                Self::I16(ConstantDataU16::try_into(data).unwrap())
+            }
+            types::I32 => {
+                let data = p.match_constant_data(ty).unwrap();
+                Self::I32(ConstantDataU32::try_into(data).unwrap())
+            }
+            types::I64 => {
+                let data = p.match_constant_data(ty).unwrap();
```

Comment:
It seems we could avoid using ConstantData for types that aren't SIMD, and have something more straightforward avoiding al the ConstantDataU{...} variants.

Reply:
I have removed ConstantData use for I8, I32 and I64.